applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h

```diff
@@ -110,40 +112,243 @@ class MapperVertexMorphingMatrixFree : public Mapper
+        mXValuesDestination.clear();
+
+        // Perform mapping
+        for(auto& node_i : mrDestinationMdpa.Nodes())
+        {
+            NodeVector neighbor_nodes(mMaxNumberOfNeighbors);
+            std::vector<double> resulting_squared_distances(mMaxNumberOfNeighbors);
+            unsigned int number_of_neighbors = mpSearchTree->SearchInRadius( node_i,
+                                                                             mFilterRadius,
+                                                                             neighbor_nodes.begin(),
+                                                                             resulting_squared_distances.begin(),
+                                                                             mMaxNumberOfNeighbors );
 
-        CreateSearchTreeIfNecessary();
-        ClearVectorsForMappingToGeometrySpace();
-        MapVariableComponentwiseToGeometrySpace( rNodalVariable );
-        AssignResultingGeometryVectorsToNodalVariable( rNodalVariableInGeometrySpace );
+            ThrowWarningIfNumberOfNeighborsExceedsLimit(node_i, number_of_neighbors);
 
-        std::cout << "> Time needed for mapping: " << mapping_time.ElapsedSeconds() << " s" << std::endl;
+            std::vector<double> list_of_weights( number_of_neighbors, 0.0 );
+            double sum_of_weights = 0.0;
+            ComputeWeightForAllNeighbors( node_i, neighbor_nodes, number_of_neighbors, list_of_weights, sum_of_weights );
+
+            int node_i_mapping_id = node_i.GetValue(MAPPING_ID);
+            for(unsigned int neighbor_itr = 0 ; neighbor_itr<number_of_neighbors ; neighbor_itr++)
+            {
+                double weight = list_of_weights[neighbor_itr] / sum_of_weights;
+                ModelPart::NodeType& node_j = *neighbor_nodes[neighbor_itr];
+
+                mXValuesDestination[node_i_mapping_id] += weight*node_j.FastGetSolutionStepValue(rVariable);
+            }
+        }
+
+        // Assign results to nodal variable
+        for(auto& node_i : mrDestinationMdpa.Nodes())
+        {
+            int i = node_i.GetValue(MAPPING_ID);
+            node_i.FastGetSolutionStepValue(rMappedVariable) = mXValuesDestination[i];
+        }
+
+        std::cout << "> Finished mapping in " << mapping_time.ElapsedSeconds() << " s." << std::endl;
+    }
+
+    // --------------------------------------------------------------------------
+    void InverseMap( const Variable<array_3d> &rVariable, const Variable<array_3d> &rMappedVariable ) override
+    {
+        if (mIsMappingInitialized == false)
+            Initialize();
+
+        BuiltinTimer mapping_time;
+        std::cout << "\n> Starting inverse mapping of " << rVariable.Name() << "..." << std::endl;
+
+        // Prepare vectors for mapping
+        mXValuesOrigin.clear();
+        mYValuesOrigin.clear();
+        mZValuesOrigin.clear();
+
+        // Perform mapping
+        for(auto& node_i : mrDestinationMdpa.Nodes())
```

Comment:
This should be 
```
        for(auto& node_i : mrOriginMdpa.Nodes())
```
and the node search should be done on the destination. (Does not matter if origin == destination)

Than it is the consistent inverse map.

The conservative version (transpose) is not really possible because of the normalization issue. We can discuss this in detail next week.

Reply:
sorry for the noice, i have overseen the way you are doing the multiplication:
```
                mXValuesOrigin[neighbor_node_mapping_id] += weight*nodal_variable[0];
                mYValuesOrigin[neighbor_node_mapping_id] += weight*nodal_variable[1];
                mZValuesOrigin[neighbor_node_mapping_id] += weight*nodal_variable[2];
```